### PR TITLE
Expose status of cron and retry workflow executions

### PR DESCRIPTION
**Pull request by dnr**
(Second try of #1745. It might be easier to look at the commits separately, especially to see what changed from that one.)

---

**What changed?**

Make cron and retry workflow executions show up with more meaningful status and history instead of all "continued as new".

**Why?**

To be able to easily see the status of retried executions and cron executions that fail or time out. Fixes issue #1545.

**How did you test it?**

Unit tests, changed existing integration tests, added a couple integration tests.

**Potential risks**

There may be more edge cases, and also external code that might be looking at specific event types.

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/workflow/mutable_state_impl.go [hunk 1]
```diff
@@ -4403,6 +4410,12 @@ func (_ *MutableStateImpl) unixNanoToTime(
 	return time.Unix(0, timestampNanos).UTC()
 }
 
+func (e *MutableStateImpl) setNewExecutionRunID(newExecutionRunID string) {
+	e.executionInfo.NewExecutionRunId = newExecutionRunID
+	// Report to parent if there is no new execution, i.e. no more retries or cron or continue-as-new.
+	e.executionInfo.ReportCompletionToParent = newExecutionRunID == ""
```
Comment (yiminc) [inline service/history/workflow/mutable_state_impl.go:4416]: I prefer not to have this extra field and rely on NewExecutionRunId to tell if you need to report to parent or not.

**Review by yiminc — APPROVED** [review]

File: service/history/workflow/mutable_state_impl.go (hunk 1 above)
Comment (dnr) [reply service/history/workflow/mutable_state_impl.go:4416]: I agree, I'll remove it
